src/librustc_mir/transform/type_check.rs

```diff
@@ -1008,6 +1009,91 @@ impl<'a, 'gcx, 'tcx> TypeChecker<'a, 'gcx, 'tcx> {
         }
     }
 
+    fn aggregate_field_ty(&mut self, ak: &Box<AggregateKind<'tcx>>, field: usize, location: Location)
+        -> Result<Ty<'tcx>, FieldAccessError>
+    {
+        let tcx = self.tcx();
+
+        let (variant, substs) = match **ak {
+            AggregateKind::Adt(def, variant, substs, _) => { // handle unions?
+                (&def.variants[variant], substs)
+            },
+            AggregateKind::Closure(def_id, substs) => {
+                return match substs.upvar_tys(def_id, tcx).nth(field) {
+                    Some(ty) => Ok(ty),
+                    None => Err(FieldAccessError::OutOfRange {
+                        field_count: substs.upvar_tys(def_id, tcx).count()
+                    }),
+                }
+            },
+            AggregateKind::Generator(def_id, substs, _) => {
+                if let Some(ty) = substs.upvar_tys(def_id, tcx).nth(field) {
+                    return Ok(ty);
+                }
+
+                return match substs.field_tys(def_id, tcx).nth(field) {
+                    Some(ty) => Ok(ty),
+                    None => Err(FieldAccessError::OutOfRange {
+                        field_count: substs.field_tys(def_id, tcx).count() + 1
+                    }),
+                }
+            },
+            AggregateKind::Array(ty) => {
+                return Ok(ty);
+            },
+            AggregateKind::Tuple => {
+                unreachable!("This should have been covered in check_rvalues");
+            },
+        };
+
+        if let Some(field) = variant.fields.get(field) {
+            Ok(self.normalize(&field.ty(tcx, substs), location))
+        } else {
+            Err(FieldAccessError::OutOfRange { field_count: variant.fields.len() })
+        }
+    }
+
+    #[allow(dead_code)]
```

Comment:
why `dead_code`?

Reply:
I was writing `check_rvalue` before I was using it and was trying to check the types. This is cleaned up in the latest commit.